### ISSUE TITLE
Custom @XmlJavaTypeAdapter ignored when applied to Map

### ISSUE THREAD

**Issue by apuig** [FasterXML/jackson-dataformat-xml#37]
We need(_) a XmlAdapter because the XML element name will use the map key (containing invalid characters(_*))

``` java
@XmlJavaTypeAdapter(URLEncoderMapDataAdapter.class)
Map<String,String> map;
...

map.put("my/key","myvalue")
```

``` xml
<my/key>my/value</my/key>)
```

(_) it's not a good idea to use schema-free attributes in a DTO, and JAXB solution looks ad-hoc and i'm not sure someone will have some real use case
(_*) one can take care (encode) before setting an URL as key

This simple test shows two different errors for 

``` java
@XmlJavaTypeAdapter(URLEncoderMapDataAdapter.class)
```

or

``` java
@XmlJavaTypeAdapter(value = URLEncoderMapDataAdapter.class, type = MapData.class)
```

``` java
package com.fasterxml.jackson.dataformat.xml.jaxb;

import static java.util.Collections.singletonMap;

import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import java.net.URLEncoder;
import java.util.HashMap;
import java.util.Map;

import javax.xml.bind.annotation.XmlRootElement;
import javax.xml.bind.annotation.adapters.XmlAdapter;
import javax.xml.bind.annotation.adapters.XmlJavaTypeAdapter;

import com.fasterxml.jackson.databind.AnnotationIntrospector;
import com.fasterxml.jackson.databind.introspect.JacksonAnnotationIntrospector;
import com.fasterxml.jackson.dataformat.xml.XmlAnnotationIntrospector;
import com.fasterxml.jackson.dataformat.xml.XmlMapper;
import com.fasterxml.jackson.dataformat.xml.XmlTestBase;

/**
 * Tring to use a custom XmlJavaTypeAdapter in order to modify how Maps are serialized
 */
public class TestXmlJavaTypeAdapter extends XmlTestBase
{
    /*
    /**********************************************************************
    /* Helper types
    /**********************************************************************
     */

    @XmlJavaTypeAdapter(URLEncoderMapDataAdapter.class)
    public static class MapData
    {
        public String key;

        public String value;

        public MapData()
        {

        }

        public MapData(String key, String value)
        {
            super();
            this.key = key;
            this.value = value;
        }
    }

    public static class URLEncoderMapDataAdapter extends XmlAdapter<MapData[], Map<String, String>>
    {
        public URLEncoderMapDataAdapter()
        {

        }

        public MapData[] marshal(Map<String, String> arg0) throws Exception
        {
            MapData[] mapElements = new MapData[arg0.size()];
            int i = 0;
            for (Map.Entry<String, String> entry : arg0.entrySet())
            {
                mapElements[i++] = new MapData(encodeKey(entry.getKey()), entry.getValue());
            }

            return mapElements;
        }

        public Map<String, String> unmarshal(MapData[] arg0) throws Exception
        {
            Map<String, String> r = new HashMap<String, String>();
            for (MapData mapelement : arg0)
            {
                r.put(decodeKey(mapelement.key), mapelement.value);
            }
            return r;
        }

        private final static String ENCODING = "UTF-8";

        private String encodeKey(String key) throws UnsupportedEncodingException
        {
            return URLEncoder.encode(key, ENCODING);
        }

        private String decodeKey(String key) throws UnsupportedEncodingException
        {
            return URLDecoder.decode(key, ENCODING);
        }
    }

    @XmlRootElement(name = "DocWithMapData")
    public static class DocWithMapData
    {
        @XmlJavaTypeAdapter(value = URLEncoderMapDataAdapter.class, type = MapData.class)
        public Map<String, String> mapDatas;
    }

    @XmlRootElement(name = "DocWithMapDataSimpleAnnotation")
    public static class DocWithMapDataSimpleAnnotation
    {
        @XmlJavaTypeAdapter(URLEncoderMapDataAdapter.class)
        public Map<String, String> mapDatas;
    }

    private Map<String, String> simpleMapData = singletonMap("key", "value");

    private Map<String, String> needEncodingMapData = singletonMap("my/key", "my/value");

    /*
     * /********************************************************************** /* Set up
     * /**********************************************************************
     */

    protected XmlMapper _jaxbMapper;

    protected XmlMapper _nonJaxbMapper;

    @Override
    public void setUp() throws Exception
    {
        super.setUp();
        _jaxbMapper = new XmlMapper();
        _nonJaxbMapper = new XmlMapper();
        // Use JAXB-then-Jackson annotation introspector
        AnnotationIntrospector intr =
            XmlAnnotationIntrospector.Pair.instance(new XmlJaxbAnnotationIntrospector(),
                new JacksonAnnotationIntrospector());
        _jaxbMapper.setAnnotationIntrospector(intr);
    }

    /*
     * /********************************************************************** /* Unit tests
     * /**********************************************************************
     */

    public void testSimpleKeyMap() throws Exception
    {
        DocWithMapData bean = new DocWithMapData();
        bean.mapDatas = simpleMapData;

        assertEquals("<DocWithMapData><mapDatas><key>value</key></mapDatas></DocWithMapData>",
            _jaxbMapper.writeValueAsString(bean));
    }

    public void testNeedEncodingKeyMap() throws Exception
    {
        DocWithMapData bean = new DocWithMapData();
        bean.mapDatas = needEncodingMapData;

        assertEquals(
            "<DocWithMapData><mapDatas><my%2Fkey>my/value</my%2Fkey></mapDatas></DocWithMapData>",
            _jaxbMapper.writeValueAsString(bean));
    }

    public void testSimpleKeyMapSimpleAnnotation() throws Exception
    {
        DocWithMapDataSimpleAnnotation bean = new DocWithMapDataSimpleAnnotation();
        bean.mapDatas = simpleMapData;

        assertEquals(
            "<DocWithMapDataSimpleAnnotation><mapDatas><key>value</key></mapDatas></DocWithMapDataSimpleAnnotation>",
            _jaxbMapper.writeValueAsString(bean));
    }

    public void testNeedEncodingKeyMapSimpleAnnotation() throws Exception
    {
        DocWithMapDataSimpleAnnotation bean = new DocWithMapDataSimpleAnnotation();
        bean.mapDatas = needEncodingMapData;

        assertEquals(
            "<DocWithMapDataSimpleAnnotation><mapDatas><my%2Fkey>my/value</my%2Fkey></mapDatas></DocWithMapDataSimpleAnnotation>",
            _jaxbMapper.writeValueAsString(bean));
    }

    public void testNeedEncodingKeyMap_nonJaxb() throws Exception
    {
        DocWithMapData bean = new DocWithMapData();
        bean.mapDatas = needEncodingMapData;

        assertEquals(
            "<DocWithMapData><mapDatas><my%2Fkey>my/value</my%2Fkey></mapDatas></DocWithMapData>",
            _nonJaxbMapper.writeValueAsString(bean));
    }

    public void testNeedEncodingKeyMapSimpleAnnotation_nonJaxb() throws Exception
    {
        DocWithMapDataSimpleAnnotation bean = new DocWithMapDataSimpleAnnotation();
        bean.mapDatas = needEncodingMapData;

        assertEquals(
            "<DocWithMapDataSimpleAnnotation><mapDatas><my%2Fkey>my/value</my%2Fkey></mapDatas></DocWithMapDataSimpleAnnotation>",
            _nonJaxbMapper.writeValueAsString(bean));
    }
}

```

**Comment by meifans** [comment]
I come across same problem. But in my situation,without @XmlJavaTypeAdapter will work as expect，on the contrary adding it will out of control . I be looking for some ways to let jackson ignore the   @XmlJavaTypeAdapter. So ,Is there anyone here?

**Comment by cowtowncoder** [comment]
@meifans One thing to note is that since your problem is not the same as one reported here, it is usually better to file a new issue. Otherwise issues tend to become hard to follow when all kinds of things that may or may not be related are clumped together.
But I'll answer this here for now; if more info needed, please file a new issue.

One way you can achieve this is by either disabling JAXB annotation processing (just don't register it), or by sub-classing `JAXBAnnotationIntrospector` and overriding method(s) that accesses this annotation. I do not remember off-hand which one that is, but should be easy enough to see from code.

Another possibility might be to use mix-in annotation to override `@XmlJavaTypeAdapter`, to specify something else, such as adapter that does not actually do anything.

**Comment by meifans** [comment]
@cowtowncoder Thanks, I got through the problem.

**Comment by cowtowncoder** [comment]
I think that the answer is that one has to use Jackson's own key serializer functionality; JAXB adapters will not work. Closing as I have no plans on working on adapter support at this point.

**Comment by cowtowncoder** [comment]
Re-opening briefly to investigate with Claude.

**Comment by cowtowncoder** [comment]
Ok: cannot be supported: closing again.
